### PR TITLE
Add note about type mismatch in automatically inserted apply argument

### LINKED ISSUE #19680: strange error for missing `using` keyword in implicit argument
## Compiler version

3.3.1  and 3.4.0-RC4.

## Minimized example

```Scala
@main
def main(args: String*): Unit =
  println("hello")
  val config = Config(8080, "docs.scala-lang.org")
  given Config = config
  println(renderWebsite("/home")(config))

case class Config(port: Int, baseUrl: String)

def renderWebsite(path: String)(using config: Config): String =
  "<html>" + renderWidget(List("cart")) + "</html>"

def renderWidget(items: List[String])(using config: Config): String =
  "<div>" + items.mkString(
    ", "
  ) + s" ${config.port} ${config.baseUrl}" + "</div>"

```

## Output Error/Warning message

```scala
-- [E007] Type Mismatch Error: /private/tmp/a.scala:6:33 -----------------------
6 |  println(renderWebsite("/home")(config))
  |                                 ^^^^^^
  |                                 Found:    (config : Config)
  |                                 Required: Int
  |
  | longer explanation available when compiling with `-explain`
1 error found
Errors encountered during compilation
```

## Why this Error/Warning was not helpful

Obviously the error message is wrong.

## Suggested improvement

It should remind me removing the whole argument or adding keyword `using` in line 6.

### PR DESCRIPTION
**Pull request by mbovel**
Started during the last spree with @jan-pieter and @iusildra.

Fixes #19680.

### CODE REVIEW

**Comment by mbovel** [comment]
This not ready yet, there would be subtleties to take care off like correctly wrapping the error message, but I am already adding @odersky as a reviewer for a high-level review: is the approach reasonable?

**Comment by mbovel** [comment]
The note is not always relevant, for example consider:

```scala
def Test = List(1,2)("hello")
```

Does that clarify anything to say that:

```
  | The required type comes from a parameter of the automatically
  | inserted `apply` method of `List[Int]`,
  | which is the type of `List.apply[Int]([1,2 : Int]*)`.
```
?

As a middle ground, I moved the message to `-explain`.

File: compiler/src/dotty/tools/dotc/typer/Applications.scala [hunk 1]
```diff
@@ -1084,7 +1084,37 @@ trait Applications extends Compatibility {
...
+
+                /** If the applied function is an automatically inserted `apply`
+                  * method and one of its arguments has a type mismatch , append
+                  * a note to the error message that explains where the required
+                  * type comes from. See #19680 and associated test case.
+                  */
+                def insertedApplyNote() =
+                  if fun1.symbol.name == nme.apply && fun1.span.isSynthetic then
+                    fun1 match
+                      case Select(qualifier, _) =>
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/Applications.scala:1100]: Suggestion: `fun1 @ Select(qualifier, _) =>`

File: compiler/src/dotty/tools/dotc/typer/Applications.scala [hunk 2]
```diff
@@ -1084,7 +1084,37 @@ trait Applications extends Compatibility {
...
+                        failedState.reporter.mapBufferedMessages:
+                          case dia: Diagnostic.Error =>
+                            dia.msg match
+                              case msg: TypeMismatch =>
+                                msg.inTree match
+                                  case Some(arg) if tree.args.exists(_.span == arg.span) =>
+                                    val Select(qualifier, _) = fun1: @unchecked
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/Applications.scala:1107]: Then this line can be dropped.

File: compiler/src/dotty/tools/dotc/typer/Applications.scala [hunk 3]
```diff
@@ -1084,7 +1084,37 @@ trait Applications extends Compatibility {
...
+                                    val noteText =
+                                      i"""The required type comes from a parameter of the automatically
+                                         |inserted `apply` method of `${qualifier.tpe}`,
+                                         |which is the type of `${qualifier.show}`.""".stripMargin
+                                    Diagnostic.Error(msg.appendExplanation("\n\n" + noteText), dia.pos)
+                                  case _ => dia
+                              case msg => dia
+                          case dia => dia
+                      case _ => ()
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/Applications.scala:1116]: Should this not be `dia`? The way it is written the return type of the method is `Diagnostic | Unit`.

**Review by odersky — CHANGES_REQUESTED** [review]
Nice improvement. Otherwise LGTM

File: compiler/src/dotty/tools/dotc/typer/Applications.scala (hunk 2 above)
Comment (mbovel) [reply compiler/src/dotty/tools/dotc/typer/Applications.scala]: This line can actually be dropped even without the previous suggestion; this is a left-over.

File: compiler/src/dotty/tools/dotc/typer/Applications.scala (hunk 3 above)
Comment (mbovel) [reply compiler/src/dotty/tools/dotc/typer/Applications.scala]: `insertedApplyNote` returns `Unit`, while the lambda passed to `mapBufferedMessages` returns `Diagnostic`, so this was correct. I named the lambda and added explicit types to make it clearer.

**Comment by mbovel** [comment]
> The pickling failure in "Applications.scala" is just due to an extra line break…
> 
> ```diff
> ➜  ~/dotty git:(mb/19680) diff before-pickling.txt after-pickling.txt
> 17956,17957c17956,17957
> <                                                                                             dotty.tools.dotc.reporting.Diagnostic
> <                                                                                               .Error
> ---
> >                                                                                             dotty.tools.dotc.reporting.
> >                                                                                               Diagnostic.Error
> ```

I was able to reproduce (with `scala3-compiler-bootstrapped / testOnly *BootstrappedOnlyCompilationTests -- *picklingWithCompiler`), but not to minimize and find exactly where and why the line breaking mechanism differs before and after pickling.

<details>

<summary>What I tried</summary>

I tried this, but it compiles successfully with `-Xprint-types -Ytest-pickler -Yprint-pos -Yprint-pos-syms`:

```scala
package dotty.tools.dotc.reporting.Diagnostic
package Diagnostic

class Error

def adipiscing() =
  def elit() =
    def nunc() =
      def vel() =
        def sapien() =
          def congue() =
            def lacus() =
              def duis() =
                def quam() =
                  def eleifend() =
                    def ultricies() =
                      def morbi() =
                        def pellentesque() =
                          val dia: Any = ???
                          dia match
                            case _: String =>
                              dia match
                              case _: String =>
                                dia match
                                  case _: String => new Diagnostic.Error()
                                  case _ => dia
                              case _ => dia
                            case _ => dia
                        pellentesque()
                      morbi()
                    ultricies()
                  eleifend()
                quam()
              duis()
            lacus()
          congue()
        sapien()
      vel()
    nunc()
  elit()

@main def Test = adipiscing()
``` 

</details>

As a workaround, I just moved `maybeAddInsertedApplyNote` to the outer scope, which avoids the problem.

**Comment by mbovel** [comment]
@odersky: I removed the last line of the message, as discussed.

**Review by odersky — APPROVED** [review]

**Pull request merged**